docs/3-cloud-computing/3.2.5-lambda.md

```diff
@@ -0,0 +1,125 @@
+      bucket = event['Records'][0]['s3']['bucket']['name']
+      key = event['Records'][0]['s3']['object']['key']
+      s3_object = s3_resource.Object(bucket,key)
+      
+      data = s3_object.get()['Body'].read().decode("utf-8").splitlines()
+      lines=csv.reader(data)
+      headers=next(lines)
+      
+      for user_data in lines:
+         print(user_data)
+
+         # Create table items here
+      
+   except Exception as e:
+      print(str(e))
+            
+   return {
+      'statusCode': 200,
+      'body': json.dumps('CSV success')
+   }
+
+```
+
+4. Once your python code is ready, download this csv file [zillow.csv](img3/zillow.csv)
+
+5. Upload the csv file to the bucket you created at the beginning of the exercise.
+
+6. After uploading the file, the lambda function should run. if you're having issues make sure to check the logs (cloud watch), and go over your code.
+
+## Exercise - Creating a Lambda function to spin up ec2 instances
+
+1. Create a Lambda function and verify it was created in the aws console.
+2. Next we want to create a DynamoDB database and table.  To do this go to the DynamoDB service page and click `"create table"`.  Create an appropriate name and make the partition key something that will denote different table entries such as "index" or "instance".
+3. Create an EC2 instance and tag it with Key: `test-<your_name>-lambda` Value: `True`  or something else that distinguishes your instance from any others.
+4. In your Lambda function create:
+ - An EC2 client
+ - A Dynamodb resource
+ - A Table connected to a specific table you have created in that
+   Dynamodb resource.
+
+``` python
+import json
+import boto3
+import csv
+
+def lambda_handler(event, context):
+   
+   try: 
+      region = 'us-east-1'
+      ec2 = boto3.client('ec2', region_name=region)
+      dynamodb = boto3.resource('dynamodb')
+      table = dynamodb.Table('<your_table>')
+      
+     
+   except Exception as e:
+      print(str(e))
+            
+   return {
+      'statusCode': 200,
+      'body': json.dumps('It got to the end of the function')
+    }
+```
+
+7. Collect and print out the data on the instance we created earlier to see the datas structure as well as some of the available information.
+
+?> One command we can use to grab all of our EC2 instances is describe_instances() with a Filter passed in.
+
+```
+filter = [{'Name': 'tag:<your_tag>', 'Values':['<your_value>']}]
+instances = ec2 describe_instances(Filters=filter)
+```
+
+### Deep Dive
+
+
+Now for something to try on your own.  We want to be able to stop our instance based on the time we last started it. The field for this is "Launch Time".  The command that gives us access to this is describe_instance_status(). Pick a time and test the code to see if our instance is correctly stopped and also store this data along with a few other things you feel we should know about this instance or instances inside our Dynamodb database.
```

Comment:
This paragraph feels repetitive with the numbered section, could reword it, or cut down on either the paragraph or the numbered list.